### ISSUE TITLE
Unable to Compile Driver

### ISSUE THREAD

**Issue by erwinsetiawan** [mongodb/mongo-php-driver#883]
### Description

Briefly describe the issue. Feel free to cross-reference your report with
another related GitHub or JIRA issue for additional context.

Before reporting an issue, check that you are using the most recent version of
the extension.

### Environment

What operating system and PHP version are you using? Mention if PHP was
installed through a third-party bundle (e.g. MAMP, XAMPP) and report its version
if applicable.

Include version information for the extension, libmongoc, and libbson. This may
be collected by grepping `phpinfo()` output like so:

```
$ php -i | grep -E 'mongodb|libmongoc|libbson'
```

In some cases, it may be helpful to provide all `phpinfo()` output from `php -i`
and a list of all install modules with `php -m`.

### Test Script

If possible, attach a complete PHP script that can be excuted on its own to
reproduce the issue. Clarify whether this script can be run from the CLI or if
it can only be reproduced through a web SAPI.

```
<?php

// TEST SCRIPT
```

If the error only presents itself in a web SAPI, report whether the issue can be
reproduced with PHP's built-in web server, which may be launched with
`php -S localhost:8000`. If not, please note your web server version in the
environment section.

http://php.net/manual/en/features.commandline.webserver.php

### Expected and Actual Behavior

How did the script fail to perform the expected task? Feel free to include debug
and/or `var_dump()` output.

If the observed behavior is an unexpected exception, please include its full
message and a backtrace (if relevant). If you encountered a segmentation fault,
please include a GDB backtrace.

https://bugs.php.net/bugs-generating-backtrace.php

```
DEBUG OR BACKTRACE
```

### Debug Log

If the issue relates to internal driver behavior (e.g. connection issues),
please include a debug log. This may be generated by setting the `mongodb.debug`
INI option to "stderr" or a directory (useful for web SAPIs).

http://php.net/manual/en/mongodb.configuration.php#ini.mongodb.debug

```
DEBUG LOG
```
PHP XAMPP 7.2.7
on mac

Show this error when running make all

tmp/mongo-php-driver/src/libmongoc/src/libmongoc/src/mongoc/mongoc-scram.c:85:10: fatal error:
      'unicode/usprep.h' file not found
#include <unicode/usprep.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [src/libmongoc/src/libmongoc/src/mongoc/mongoc-scram.lo] Error 1

**Comment by derickr** [comment]
Can you please provide the output of the ``./configure`` stage? This file is part of ICU, which the configure script should check for.

**Comment by jmikola** [comment]
Additionally, please report the exact version of the driver (either tagged release or commit hash) that you're attempting to install.

**Comment by erwinsetiawan** [comment]
Here the ./configure stage

`checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for a sed that does not truncate output... /usr/bin/sed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-apple-darwin17.6.0
checking host system type... x86_64-apple-darwin17.6.0
checking target system type... x86_64-apple-darwin17.6.0
checking for PHP prefix... /Applications/XAMPP/xamppfiles
checking for PHP includes... -I/Applications/XAMPP/xamppfiles/include/php -I/Applications/XAMPP/xamppfiles/include/php/main -I/Applications/XAMPP/xamppfiles/include/php/TSRM -I/Applications/XAMPP/xamppfiles/include/php/Zend -I/Applications/XAMPP/xamppfiles/include/php/ext -I/Applications/XAMPP/xamppfiles/include/php/ext/date/lib
checking for PHP extension directory... /Applications/XAMPP/xamppfiles/lib/php/extensions/no-debug-non-zts-20170718
checking for PHP installed headers prefix... /Applications/XAMPP/xamppfiles/include/php
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... no
checking for awk... awk
checking if awk is broken... no
checking whether to enable MongoDB support... yes, shared
checking PHP version... 7.2.7
checking whether to enable developer build flags... no
checking whether to enable code coverage... no
checking whether to use system libbson... no
checking whether to use system libmongoc... no
checking for gcc... (cached) cc
checking whether we are using the GNU C compiler... (cached) yes
checking whether cc accepts -g... (cached) yes
checking for cc option to accept ISO C89... (cached) none needed
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking accept ARG2 => struct sockaddr ARG3 => socklen_t ... ok
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for typeof syntax and keyword spelling... typeof
checking for __sync_add_and_fetch_4... yes
checking for __sync_add_and_fetch_8... yes
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for stdbool.h that conforms to C99... yes
checking for _Bool... yes
checking for strings.h... (cached) yes
checking whether byte ordering is bigendian... no
checking for strnlen... yes
checking for reallocf... yes
checking for syscall... yes
checking for snprintf... yes
checking for struct timespec... yes
checking for library containing clock_gettime... none required
checking for library containing floor... none required
checking for gmtime_r... yes
checking for rand_r... yes
checking if compiler needs -Werror to reject unknown flags... yes
checking for the pthreads library -lpthreads... no
checking whether pthreads work without any flags... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... -D_THREAD_SAFE
checking for PTHREAD_PRIO_INHERIT... yes
checking whether PTHREAD_ONCE_INIT needs braces... no
checking for pkg-config... /Applications/XAMPP/xamppfiles/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for PHP_MONGODB_SNAPPY... no
checking for snappy_uncompress in -lsnappy... no
checking snappy-c.h usability... no
checking snappy-c.h presence... no
checking for snappy-c.h... no
checking for PHP_MONGODB_ZLIB... yes
./configure: line 10588: cd: @@BITROCK_COMMON_ROOTDIR@@: No such file or directory
checking for res_nsearch... yes
checking for res_ndestroy... yes
checking whether to enable SASL for Kerberos authentication... auto
checking for PHP_MONGODB_SASL... no
checking for sasl_client_init in -lsasl2... yes
checking sasl/sasl.h usability... yes
checking sasl/sasl.h presence... yes
checking for sasl/sasl.h... yes
checking for sasl_client_done in -lsasl2... yes
checking which SASL library to use... cyrus
checking whether to enable crypto and TLS... auto
checking deprecated option for OpenSSL library path... auto
configure: checking whether Darwin SSL is available
checking which TLS library to use... darwin
checking whether to use system crypto profile... no
checking deprecated option for whether to use system crypto profile... no
checking whether to enable ICU for SASLPrep with SCRAM-SHA-256 authentication... auto
checking for PHP_MONGODB_ICU... yes
./configure: line 14035: cd: @@BITROCK_COMMON_ROOTDIR@@: No such file or directory
checking for shm_open... yes
checking for sched_getcpu... no
checking for socklen_t... yes
checking for pthread_join in LIBS= with CFLAGS=-D_THREAD_SAFE ... yes
checking if compiler needs -Werror to reject unknown flags... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... -D_THREAD_SAFE
checking for PTHREAD_PRIO_INHERIT... (cached) yes
checking whether C compiler accepts -Wno-deprecated-declarations... yes
checking whether C compiler accepts -Wno-cast-align... yes
checking whether C compiler accepts -Wno-unneeded-internal-declaration... yes
checking whether C compiler accepts -Wno-error=unused-command-line-argument... yes
checking if weak symbols are supported... yes
checking for ld used by cc... /Library/Developer/CommandLineTools/usr/bin/ld
checking if the linker (/Library/Developer/CommandLineTools/usr/bin/ld) is GNU ld... no
checking for /Library/Developer/CommandLineTools/usr/bin/ld option to reload object files... -r
checking for BSD-compatible nm... /usr/bin/nm -B
checking whether ln -s works... yes
checking how to recognize dependent libraries... pass_all
checking dlfcn.h usability... yes
checking dlfcn.h presence... yes
checking for dlfcn.h... yes
checking the maximum length of command line arguments... 196608
checking command to parse /usr/bin/nm -B output from cc object... ok
checking for objdir... .libs
checking for ar... ar
checking for ranlib... ranlib
checking for strip... strip
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking if cc supports -fno-rtti -fno-exceptions... yes
checking for cc option to produce PIC... -fno-common
checking if cc PIC flag -fno-common works... yes
checking if cc static flag -static works... no
checking if cc supports -c -o file.o... yes
checking whether the cc linker (/Library/Developer/CommandLineTools/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin17.6.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no

creating libtool
appending configuration tag "CXX" to libtool
checking for ld used by g++... /Library/Developer/CommandLineTools/usr/bin/ld
checking if the linker (/Library/Developer/CommandLineTools/usr/bin/ld) is GNU ld... no
checking whether the g++ linker (/Library/Developer/CommandLineTools/usr/bin/ld) supports shared libraries... yes
checking for g++ option to produce PIC... -fno-common
checking if g++ PIC flag -fno-common works... yes
checking if g++ static flag -static works... no
checking if g++ supports -c -o file.o... yes
checking whether the g++ linker (/Library/Developer/CommandLineTools/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin17.6.0 dyld
(cached) (cached) checking how to hardcode library paths into programs... immediate
configure: creating ./config.status

mongodb was configured with the following options:

Build configuration:
  CFLAGS                                           : -g -O2
  Extra CFLAGS                                     :  -Wno-deprecated-declarations -Wno-cast-align -Wno-unneeded-internal-declaration -Wno-error=unused-command-line-argument
  Developers flags (slow)                          :
  Code Coverage flags (extra slow)                 :
  System mongoc                                    : no
  System libbson                                   : no
  LDFLAGS                                          :
  EXTRA_LDFLAGS                                    :
  MONGODB_SHARED_LIBADD                            : -Wl,-rpath,/lib -L/lib -framework Security -framework CoreFoundation -lsasl2 -lresolv -Wl,-rpath,/lib -L/lib  -lz -licuuc -licudata

Please submit bugreports at:
  https://jira.mongodb.org/browse/PHPC


config.status: creating /Users/erwinsetiawan/works/tmp/mongo-php-driver/src/libmongoc/src/libbson/src/bson/bson-config.h
config.status: creating /Users/erwinsetiawan/works/tmp/mongo-php-driver/src/libmongoc/src/libbson/src/bson/bson-version.h
config.status: creating /Users/erwinsetiawan/works/tmp/mongo-php-driver/src/libmongoc/src/libmongoc/src/mongoc/mongoc-config.h
config.status: creating /Users/erwinsetiawan/works/tmp/mongo-php-driver/src/libmongoc/src/libmongoc/src/mongoc/mongoc-version.h
config.status: creating config.h
config.status: config.h is unchanged`

and I using the latest version of driver 1.5.1

Thank You

**Comment by jmikola** [comment]
Can you share the output for the following?

```
$ pkg-config --cflags icu-uc
```

Based on the directory reported by that command (after the `-I`), can you check whether the `unicode/usprep.h` exists within it? If only that file is missing, check if a `unicode/` directory exists and share its contents.

**Comment by erwinsetiawan** [comment]
When i running 

`$ pkg-config --cflags icu-uc`

It will have output

`-I@@BITROCK_COMMON_ROOTDIR@@/include`

What I should I do?

**Comment by erwinsetiawan** [comment]
Ah i got it

$ brew install icu4c
$ brew link icu4c --force

it solved my problem, thank you for some clue.

**Comment by erwinsetiawan** [comment]
I have success compile but it can't installed, already change php.ini and add extension="mongodb.so"
restarting my Apache server and it didn't appear on phpinfo.

**Comment by derickr** [comment]
Usually there are multiple php.ini files on an OSX machines. For example, one for the Apple installed Apache/PHP, and one for the one installed through brew. It's possible that you might have updated the wrong php.ini file. If you create a PHP script with ```<?php phpinfo();``` in it, it will tell you which ``php.ini`` is in use through the Apache/PHP that you expect.

**Comment by erwinsetiawan** [comment]
Here from phpinfo()


```
Server API | Apache 2.0 Handler
-- | --
Virtual Directory Support | disabled
Configuration File (php.ini) Path | /Applications/XAMPP/xamppfiles/etc
Loaded Configuration File | /Applications/XAMPP/xamppfiles/etc/php.ini
Scan this dir for additional .ini files | (none)
```

here from php -i 
```
Server API => Command Line Interface
Virtual Directory Support => disabled
Configuration File (php.ini) Path => /Applications/XAMPP/xamppfiles/etc
Loaded Configuration File => /Applications/XAMPP/xamppfiles/etc/php.ini
Scan this dir for additional .ini files => (none)
Additional .ini files parsed => (none)
```

here from php --ini
```
Configuration File (php.ini) Path: /Applications/XAMPP/xamppfiles/etc
Loaded Configuration File:         /Applications/XAMPP/xamppfiles/etc/php.ini
Scan for additional .ini files in: (none)
Additional .ini files parsed:      (none)

```

I think it not load properly

When I running on php -i its shows mongodb driver but nothing in phpinfo();

```
php -i | grep mongodb
mongodb
```

but when I run some script inside test.php

```
<?php

$connection_string = "mongodb://server:111111";
$client = new MongoDB\Client($connection_string);

?>
```

And running on cli, Mongodb client not found

```
php test.php
PHP Fatal error:  Uncaught Error: Class 'MongoDB\Client' not found in /Applications/XAMPP/xamppfiles/htdocs/test.php:4
Stack trace:
#0 {main}
  thrown in /Applications/XAMPP/xamppfiles/htdocs/test.php on line 4

Fatal error: Uncaught Error: Class 'MongoDB\Client' not found in /Applications/XAMPP/xamppfiles/htdocs/test.php:4
Stack trace:
#0 {main}
  thrown in /Applications/XAMPP/xamppfiles/htdocs/test.php on line 4

```

**Comment by derickr** [comment]
``MongoDB\Client`` comes from the MongoDB PHP *Library* which you need to install from composer: http://php.net/manual/en/mongodb.tutorial.library.php

**Comment by erwinsetiawan** [comment]
Ah my bad forget the composer, anyway ignore the test script. I have already check and make sure it have sama php.ini file on each. Guy on the other thread suggest me to using mamp instead using xampp. So probably this issue is on xampp for mac os, not sure about this, but it works for another php ext in the same way.

**Comment by jmikola** [comment]
> Guy on the other thread suggest me to using mamp instead using xampp. So probably this issue is on xampp for mac os, not sure about this, but it works for another php ext in the same way.

I saw the comment but it looks like FUD. There are plenty of folks using the extension with XAMPP.

If `extension=mongodb.so` is listed in the INI file reported by `phpinfo()` when called from the web SAPI for XAMPP, you may just need to restart Apache (assuming it PHP is loaded as an Apache module) to ensure that the INI file is processed.

If the extension actually fails to load when the INI file is processed, then you should expect to see some PHP error/warning alerting you to that. If there is no such message _and_ the extension still fails to load, then I would double-check the error reporting settings for PHP to ensure messages are not being suppressed.

**Comment by erwinsetiawan** [comment]
@jmikola I understand exactly what do you think, I also don't why it happen on xampp. I have pretty sure double check the configuration with the same php.ini file and I also seen any error reporting on xampp. It never happen before with my previous installation with old xampp and mongodb driver version. Need to have experience by yourself then you will know. 

Finally I give up using xampp and I just using MAMP now and it seem working perfect without any issue. 

Thank you for your great work and your team also. Cheers!

**Comment by jmikola** [comment]
Glad MAMP resolved things for you. I've managed to borrow a freshly formatted MacBook Pro and am in the process of upgrading it from Sierra to High Sierra so I can test the latest version of [XAMPP](https://www.apachefriends.org/index.html). I'll carefully document the steps I take along the way and follow up on this ticket with my findings.

**Comment by jmikola** [comment]
I started with a new macOS installation of High Sierra (version 10.13.6) and downloaded [xampp-osx-7.2.7-0-r01-installer.dmg](https://www.apachefriends.org/xampp-files/7.2.7/xampp-osx-7.2.7-0-r01-installer.dmg). Following the default installation process, XAMPP was installed to `/Applications/XAMPP`. The XAMPP manager app was opened after the installation completed and "Manage Servers" showed that "Apache Web Server" running ("MySQL Database" and "ProFTPD" were stopped). The installer also opened Safari to `http://mongodbs-mackbook-pro.local/dashboard`, which I left open to return to later.

I opened a terminal and attempting to install with the MongoDB driver with `/Applications/XAMPP/bin/pecl install mongodb`:

```
$ /Applications/XAMPP/bin/pecl install mongodb
No releases available for package "pecl.php.net/mongodb"
install failed
```

I attempted to update channels:

```
$ /Applications/XAMPP/bin/pecl update-channels
channel-add: temp_dir is not writable: "/var/folders/..." - You can change this location with "pear config-set temp_dir"
```

I attempted to update the temp dir:

```
$ /Applications/XAMPP/bin/peear config-set temp_dir /tmp
config-set succeeded
```

Another attempt to update channels:

```
$ /Applications/XAMPP/bin/pecl update-channels pecl.php.net
Updating channel "pecl.php.net"
could not create lock file: fopen(/Applications/XAMPP/xamppfiles/lib/php/.lock): failed to open stream: Permission denied
...
```

At this point, I decided to use `sudo` to update channels and install the driver:

```
$ sudo /Applications/XAMPP/bin/pecl update-channels pecl.php.net
Password: 
Updating channel "pecl.php.net"
Update of Channel "pecl.php.net" succeeded
```

Attempting to install again:

```
$ sudo /Applications/XAMPP/bin/pecl install mongodb
No releases available for package "pecl.php.net/mongodb"
install failed
```

Attempting to list packages:

```
$ sudo /Applications/XAMPP/bin/pecl list-all
Connection to `ssl://pecl.php.net:443' failed:
```

At this point, I gather that there may be some incompatibility between the build of PHP included with XAMPP and the OpenSSL library on macOS High Sierra. I'll note the version of OpenSSL that PHP was compiled against and the version of OpenSSL available on macOS:

```
$ sudo /Applications/XAMPP/bin/php -i | grep "SSL Version"
SSL Version => OpenSSL/1.0.2o
$ openssl version
LibreSSL 2.2.7
```

I decide to manually download the PECL package ([mongodb-1.5.1.tgz](https://pecl.php.net/get/mongodb-1.5.1.tgz)) and install it from the local filesystem to bypass this SSL issue.

```
$ sudo /Applications/XAMPP/bin/pecl install /Users/mongodb/Downlaods/mongodb-1.5.1.tgz
423 source files, building
running: phpize

...

Cannot find autoconf. Please check your autoconf installation and the
$PHP_AUTOCONF environment variable. Then, rerun this script.

ERROR: `phpize' failed
```

I install [Homebrew](https://brew.sh/) and proceed to install the following packages:

```
$ brew install autoconf
```

I then repeat installation of the extension, which ultimately fails for the same reason cited in your [original comment](https://github.com/mongodb/mongo-php-driver/issues/883#issue-342285865), which was failure to include `unicode/usprep.h`. I notice that the `configure` output when I attempted to install the driver also includes the following suspect lines that you shared:

```
checking for pkg-config... /Applications/XAMPP/xamppfiles/bin/pkg-config

...

checking for PHP_MONGODB_ZLIB... yes
./configure: line 10588: cd: @@BITROCK_COMMON_ROOTDIR@@: No such file or directory

...

checking for PHP_MONGODB_ICU... yes
./configure: line 14035: cd: @@BITROCK_COMMON_ROOTDIR@@: No such file or directory
```

Googling for `BITROCK_COMMON_ROOTDIR` did not turn up much, but "bitrock" comes up as the developer of Bitnami, which is [affiliated with XAMPP](https://bitnami.com/stack/xampp). I also think it's suspect that XAMPP is providing its own copy of `pkg-config`, which reports `@@BITROCK_COMMON_ROOTDIR@@` in its include path (as you shared in https://github.com/mongodb/mongo-php-driver/issues/883#issuecomment-406140279). Since macOS High Sierra does not appear to ship with `pkg-config`, I decide to install it via Homebrew:

```
$ brew install pkg-config
```

If you run `configure --help` for the extension, it reports that the `PKG_CONFIG` environment variable may be assigned the "path to pkg-config utility". That is extracted from [`scripts/autotools/m4/pkg.m4`](https://github.com/mongodb/mongo-php-driver/blob/1.5.1/scripts/autotools/m4/pkg.m4), which is the build script responsible for finding `pkg-config` and using it to check for dependencies. That said, I was not successful attempting to get XAMPP's `pecl` command to respect my customized `PKG_CONFIG` environment variable. I decided to look into how the `pecl` command was picking up `/Applications/XAMPP/xamppfiles/bin/` and found that it is assigned to the `bin_dir` [PEAR configuration option](https://pear.php.net/manual/en/guide.users.commandline.config.php).


```
$ /Applications/XAMPP/bin/pecl config-get bin_dir
/Applications/XAMPP/xamppfiles/bin
```

Since this is the same directory that contains other essential binaries (e.g. `phpize`), I don't think we can alter it without introducing a new problem. I think the best solution is to override `/Applications/XAMPP/xamppfiles/bin/pkg-config` with a symlink to version installed by Homebrew.

```
$ cd /Applications/XAMPP/xamppfiles/bin
$ sudo mv pkg-config pkg-config.bak
$ sudo ln -sf `which pkg-config` pkg-config
```

Let's repeat the install process:

```
$ sudo /Applications/XAMPP/bin/pecl install /Users/mongodb/Downlaods/mongodb-1.5.1.tgz
```

The `configure` script still picks up `/Applications/XAMPP/xamppfiles/bin/pkg-config` as expected, but that is now a symlink to Homebrew's `pkg-config` binary. The `@@BITROCK_COMMON_ROOTDIR@@` errors are gone and `configure` reports that ZLIB was found but ICU was not. That's understandable as I've not installed ICU manually with Homebrew (as you did in https://github.com/mongodb/mongo-php-driver/issues/883#issuecomment-406143360). The build ultimately succeeds and instructs me to add "extnesion=mongodb.so" to `php.ini`.

```
$ /Applications/XAMPP/xamppfiles/bin/php --ini
Configuration File (php.ini) Path: /Applications/XAMPP/xamppfiles/etc
Loaded Configuration File:         /Applications/XAMPP/xamppfiles/etc/php.ini
Scan for additional .ini files in: (none)
Additional .ini files parsed:      (none)
```

I add `extension=mongodb.so` to `/Applications/XAMPP/xamppfiles/etc/php.ini` and confirm that the extension is loaded by the XAMPP's CLI environment:

```
$ /Applications/XAMPP/xamppfiles/bin/php -m | grep mongodb
mongodb
```

I return Safari, which has been open to `http://mongodbs-mackbook-pro.local/dashboard` since XAMPP was installed. There is a "PHPInfo" link at the top, which simply displays `phpinfo()` output on its own page. This reports that the PHP web SAPI is using the same INI file as the CLI environment; however, the `mongodb` extension is not listed. That is to be expected, as the Apache process and its PHP module has not been restarted. I open the XAMPP control panel, restart the "Apache Web Server", and then reload the "PHPInfo" dashboard page. The `mongodb` extension is still not listed. `phpinfo()` reports that the PHP error log is `/Applications/XAMPP/xamppfiles/logs/php_error_log, so I tail that and find:

```
[20-Jul-2018 18:47:57 UTC] PHP Warning:  PHP Startup: Unable to load dynamic library 'mongodb.so' (tried: /Applications/XAMPP/xamppfiles/lib/php/extensions/no-debug-non-zts-20170718/mongodb.so (dlopen(/Applications/XAMPP/xamppfiles/lib/php/extensions/no-debug-non-zts-20170718/mongodb.so, 9): Symbol not found: __cg_jpeg_resync_to_restart
  Referenced from: /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
  Expected in: /Applications/XAMPP/xamppfiles/lib/libjpeg.8.dylib
 in /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO), /Applications/XAMPP/xamppfiles/lib/php/extensions/no-debug-non-zts-20170718/mongodb.so.so (dlopen(/Applications/XAMPP/xamppfiles/lib/php/extensions/no-debug-non-zts-20170718/mongodb.so.so, 9): image not found)) in Unknown on line 0
```

Searching Google for this error turned up a number of forum threads referring to `DYLD_LIBRARY_PATH`. These two threads stood out:

 * https://stackoverflow.com/q/24941078/162228 involves a user attempting to executing `ffmpeg` from a PHP script using `exec()`, which demonstrates that this environment variable does not just affect the loading of PHP extensions.
 * https://stackoverflow.com/q/3146274/162228 is a more general discussion about `DYLD_LIBRARY_PATH` on macOS, and provides a lot of context about how it works. It also refers to a `DYLD_FALLBACK_LIBRARY_PATH` environment variable, but I did not see that utilized at all in XAMPP and did not pursue it.

I'll note that `DYLD_LIBRARY_PATH` actually appears as an environment variable in `phpinfo()` output with the value `/Applications/XAMPP/xamppfiles/lib/`. XAMPP appears to set `DYLD_LIBRARY_PATH` in `/Applications/XAMPP/xamppfiles/bin/envvars`. This script checks if a path is already assigned and will prepend `/Applications/XAMPP/xamppfiles/lib/` if so, or simply assign `/Applications/XAMPP/xamppfiles/bin/` is the environment variable was empty. On my fresh install of macOS High Sierra, `DYLD_LIBRARY_PATH` was not defined so XAMPP was assigning its own path. This interfers with PHP's ability to dynamically `mongodb.so` and/or the libraries it was built with. I did look defining `DYLD_LIBRARY_PATH` to `/usr/local/lib:/usr/lib` first and allowing XAMPP to prepend its own directory, but that did not help. I expect that some of the libraries shipping with XAMPP are not ABI compatible with those that ship with macOS, which were used to build `mongodb.so`.

Ultimately, I modified `/Applications/XAMPP/xamppfiles/bin/envvars` and commented out the `export DYLD_LIBRARY_PATH` line. After restarting Apache from the XAMPP control panel, I noticed that `mongodb` was now successfully loaded by the PHP web SAPI and the `DYLD_LIBRARY_PATH` environment variable was no longer present. I'm hestitant to suggest this as a final solution, as XAMPP's Apache or PHP environments may still be relying on `DYLD_LIBRARY_PATH` to load other libraries. This may be especially true for the other PHP extensions that ship with XAMPP.

In conclusion, there is definitely a conflict with XAMPP; however, I don't believe this is limited to the MongoDB driver and suspect it extends to other PECL extensions that dynamically link shared libraries. I'd argue that the best solution is to install PHP using Homebrew and manually install extensions with its `pecl` command (extension formulae no longer exist since the PHP tap was deprecated). If using XAMPP is a priority, I think the [XAMPP-VM environment](https://www.apachefriends.org/blog/new_xampp_20170628.html) is preferable to their standard macOS package, as it runs PHP inside of a Linux VM. Lastly, I'll concede that the suggestion in https://github.com/mongodb/mongo-php-driver/issues/822#issuecomment-406271673 to use MAMP is also well founded.

**Comment by erwinsetiawan** [comment]
@jmikola this is pretty great, thank you for this details. It seems can explain what happen now. 
Again, thank you very much!!

**Comment by mpp21x** [comment]
@jmikola  Thank you so much!!!

**Comment by WadeShuler** [comment]
Sorry to bump an old thread. However, this came up in a Google search while trying to fix my issue. It is similar/related and this info may help someone else who stumbles here.

NOTE: I am on MacOS Catalina and using XAMPP. This issue was present for people on High Sierra and others as well.

I was getting:

```bash
fatal error: 'unicode/usprep.h' file not found
#include <unicode/usprep.h>
         ^~~~~~~~~~~~~~~~~~
```

Notice it's complaining about `usprep.h` and make fails. While searching, this was due to `icu4c`. I had it insalled via homebrew, it was a minor version out of date.. updated it just because, no luck. hmm..

After banging my head, I ended up doing `which icu4c` which obviously failed (turns out isn't a real binary lol). I then tried `which icuinfo`, and immediately saw (hoped) that I just figured out the issue. It was pulling from my XAMPP bin directory.

I had to modify the paths to get the bin and sbin dirs for icu4c to be searched first.

More info on my SO answer here: [https://stackoverflow.com/a/67349734/2703135](https://stackoverflow.com/a/67349734/2703135)

I hope it helps someone :)
